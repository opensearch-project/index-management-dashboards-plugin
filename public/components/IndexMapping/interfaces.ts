/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import { MappingsProperties, MappingsPropertiesObject } from "../../../models/interfaces";

export type IndexMappingsAll = {
  properties?: MappingsProperties;
  [key: string]: any;
};

export type IndexMappingsObjectAll = {
  properties?: MappingsPropertiesObject;
  [key: string]: any;
};

export interface IndexMappingProps {
  value?: IndexMappingsAll;
  oldValue?: IndexMappingsAll;
  originalValue?: IndexMappingsAll;
  onChange: (value: IndexMappingProps["value"]) => void;
  isEdit?: boolean;
  oldMappingsEditable?: boolean; // in template edit case, existing mappings is editable
  readonly?: boolean;
  docVersion: string;
}

export enum EDITOR_MODE {
  JSON = "JSON",
  VISUAL = "VISUAL",
}

export interface IIndexMappingsRef {
  validate: () => Promise<string>;
  getJSONEditorValue: () => string;
}
