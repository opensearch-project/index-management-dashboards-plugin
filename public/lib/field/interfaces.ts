/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import React from "react";

// if it's a string[], the value will become nested.
// registerField({ name: ['a', 'b', 'c.d'] }) => { a: { b: { c,d: '' } } }
export type FieldName = string | string[];

export type FieldOption = {
  /**
   * All component changes will arrive here [set value will not trigger this function]
   */
  onChange?: (name: FieldName, value?: any) => void;

  /**
   * Initialization data
   */
  values?: {};

  /**
   * OriginalValues
   */
  originalValues?: {};

  unmountComponent?: boolean;
};

export type ValidateResults = {
  errors: Record<string, string[]> | null;
  values: any;
};

export type InitResult<T> = {
  value?: T;
  onChange(value: T): void;
  ref?: React.RefCallback<any>;
};

export type Rule = {
  /**
   * cannot be empty (cannot be used with pattern)
   * @default true
   */
  required?: boolean;

  /**
   * error message
   */
  message?: string;

  /**
   * Check Regular Expression
   */
  pattern?: RegExp;
  /**
   * Minimum string length /minimum number of arrays
   */
  minLength?: number;
  /**
   * Maximum string length /maximum number of arrays
   */
  maxLength?: number;

  /**
   * String exact length /array exact number
   */
  length?: number;

  /**
   * minimum
   */
  min?: number;

  /**
   * maximum value
   */
  max?: number;
  /**
   * Summary of common patterns
   */
  format?: "url" | "email" | "tel" | "number";

  /**
   * Custom verification, (don't forget to execute callback() when the verification is successful, otherwise the verification will not return)
   */
  validator?: (rule: Rule, value: string | number | object | boolean | Date | null | any) => string | Promise<string>;

  /**
   * The name of the event that triggered the validation
   */
  trigger?: "onChange" | "onBlur" | string;
};

export type InitOption = {
  /**
   * The name of the field
   */
  name: FieldName;

  /**
   * The name of the event that triggered the data change
   * @default 'onChange'
   */
  trigger?: string | "onChange" | "onBlur";

  /**
   * Check rules
   */
  rules?: Rule[];

  /**
   * Component custom events can be written here, others will be transparently transmitted (small package version ^0.3.0 support, large package ^0.7.0 support)
   */
  props?: any;
};

export type FieldInstance = {
  /**
   * Initialize each component
   */
  registerField(option?: InitOption): InitResult<any>;

  /**
   * check
   * @param name
   */
  validatePromise(name?: FieldName): Promise<ValidateResults>;

  /**
   * 	Get the value of a single input control
   * @param field name
   */
  getValue(name: FieldName): any;

  /**
   * Get the values ​​of a set of input controls, if no parameters are passed in, get the values ​​of all components
   * @param names
   */
  getValues(): any;

  /**
   * Get the values ​​of a set of input controls, if no parameters are passed in, get the values ​​of all components
   * @param names
   */
  getOriginalValues(): any;

  /**
   * Set the value of a single input control (will trigger render, please follow the timing of react)
   */
  setValue(name: FieldName, value: any): void;

  /**
   * Set the value of a set of input controls (will trigger render, please follow the timing of react)
   */
  setValues(obj: any): void;

  /**
   *
   */
  setOriginalValues(obj: any): void;

  /**
   * Reset values
   */
  resetValues(obj: any): void;

  /**
   * Delete value
   */
  deleteValue(key: FieldName): void;

  /**
   * Get the Error of a single input control
   */
  getError(name: FieldName): null | string[];

  /**
   * Get the Error for a set of input controls
   * @param names field name
   */
  getErrors(): any;

  /**
   * Sets the Error for a single input control
   * @param name
   * @param errors
   */
  setError(name: FieldName, errors: null | string[]): void;

  /**
   * Sets the Error for a set of input controls
   */
  setErrors(obj: any): void;

  /**
   * Get difference between originalValues & currentValues
   */
  computeDifference(): number;
};

export type ValidateFunction = (
  rule: Rule & { field: string; aliasName?: string },
  value: string | number | object | boolean | Date
) => string | Promise<string> | undefined;
