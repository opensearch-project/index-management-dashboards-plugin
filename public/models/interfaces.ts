/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Direction, Query } from "@elastic/eui";
import { SMPolicy } from "../../models/interfaces";
import { IJobItemMetadata } from "../lib/JobScheduler";
import {
  IndexService,
  ManagedIndexService,
  PolicyService,
  RollupService,
  TransformService,
  NotificationService,
  SnapshotManagementService,
  CommonService,
} from "../services";

export interface BrowserServices {
  indexService: IndexService;
  managedIndexService: ManagedIndexService;
  policyService: PolicyService;
  rollupService: RollupService;
  transformService: TransformService;
  notificationService: NotificationService;
  snapshotManagementService: SnapshotManagementService;
  commonService: CommonService;
}

export interface SMPoliciesQueryParams {
  from: number;
  size: number;
  sortField: keyof SMPolicy;
  sortOrder: Direction;
}

interface ArgsWithQuery {
  query: Query;
  queryText: string;
  error: null;
}
interface ArgsWithError {
  query: null;
  queryText: string;
  error: Error;
}
export type OnSearchChangeArgs = ArgsWithQuery | ArgsWithError;

export interface Toast {
  id?: string;
  title?: string;
  iconType?: string;
  color: string;
  text?: JSX.Element;
}

export interface Column {
  field: string;
  name: string;
  width: string;
  truncateText?: boolean;
  sortable: boolean;
}

export interface RestoreError {
  reason?: string;
  type?: string;
  ok?: boolean;
  error?: string;
}

export interface LatestActivities {
  activityType: "Creation" | "Deletion";
  status?: string;
  snapshot?: string;
  start_time?: number;
  end_time?: number;
  info?: {
    message?: string;
    cause?: string;
  };
}

export enum RESTORE_OPTIONS {
  restore_specific_indices = "restore_specific_indices",
  restore_all_indices = "restore_all_indices",
  do_not_rename = "do_not_rename",
  add_prefix = "add_prefix",
  rename_indices = "rename_indices",
  restore_aliases = "restore_aliases",
  include_global_state = "include_global_state",
  ignore_unavailable = "ignore_unavailable",
  partial = "partial",
  customize_index_settings = "customize_index_settings",
  ignore_index_settings = "ignore_index_settings",
}

export interface ModifiedStages {
  START: string;
  INIT: string;
  INDEX: string;
  VERIFY_INDEX: string;
  TRANSLOG: string;
  FINALIZE: string;
  DONE: string;
}

export interface IndexItem {
  index: string;
  restore_status?: string;
}

export interface ClusterInfo {
  cluster_name?: string;
}

export interface ReindexJobMetaData extends IJobItemMetadata {
  extras: {
    clusterInfo: ClusterInfo;
    toastId: string;
    sourceIndex: string;
    destIndex: string;
    writingIndex: string;
    taskId: string;
    destType: "index" | "other";
  };
}

export interface RecoveryJobMetaData extends IJobItemMetadata {
  extras: {
    clusterInfo: ClusterInfo;
    toastId: string;
    sourceIndex: string;
    destIndex: string;
    taskId: string;
  };
}

export interface ForceMergeJobMetaData extends IJobItemMetadata {
  extras: {
    clusterInfo: ClusterInfo;
    toastId: string;
    sourceIndex: string[];
    taskId: string;
  };
}

export interface OpenJobMetaData extends IJobItemMetadata {
  extras: {
    clusterInfo: ClusterInfo;
    toastId: string;
    indexes: string[];
    taskId: string;
  };
}
