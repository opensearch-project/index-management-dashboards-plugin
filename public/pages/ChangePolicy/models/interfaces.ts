/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Policy } from "../../../../models/interfaces";

export interface PolicyOption {
  label: string;
  value: Policy;
}
