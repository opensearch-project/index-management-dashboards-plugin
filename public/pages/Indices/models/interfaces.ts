/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Direction } from "@elastic/eui";
import { Policy } from "../../../../models/interfaces";
import { ManagedCatIndex } from "../../../../server/models/interfaces";

export interface PolicyOption {
  label: string;
  policy?: Policy;
}

export interface IndicesQueryParams {
  from: number;
  size: number;
  search: string;
  sortField: keyof ManagedCatIndex;
  sortDirection: Direction;
  showDataStreams: boolean;
}
