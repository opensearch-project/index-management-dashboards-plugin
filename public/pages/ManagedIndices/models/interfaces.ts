/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { Direction } from "@elastic/eui";
import { ManagedIndexItem } from "../../../../models/interfaces";

export interface ManagedIndicesQueryParams {
  from: number;
  size: number;
  search: string;
  sortField: keyof ManagedIndexItem;
  sortDirection: Direction;
  showDataStreams: boolean;
}
