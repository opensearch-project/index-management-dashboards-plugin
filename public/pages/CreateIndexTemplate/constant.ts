/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export enum TABS_ENUM {
  SUMMARY = "summary",
  CONFIG = "CONFIG",
}
export const tabs = [
  {
    id: TABS_ENUM.SUMMARY,
    name: "Summary",
  },
  {
    id: TABS_ENUM.CONFIG,
    name: "Configuration",
  },
];
