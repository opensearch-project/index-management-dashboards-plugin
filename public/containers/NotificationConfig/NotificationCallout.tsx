import { EuiCallOut, EuiLink, EuiSpacer } from "@elastic/eui";
import React from "react";
import { ROUTES } from "../../utils/constants";
import { ActionType, OperationType, OperationTypeMapTitle } from "../../pages/Notifications/constant";
import { getPermissionValue } from "./hooks";

export interface INotificationCalloutProps {
  hasDefaultNotification: boolean;
  operationType?: OperationType;
  actionType: ActionType;
  permissionForViewLRON: boolean;
  permissionForCreateLRON: boolean;
}

export default function NotificationCallout(props: INotificationCalloutProps) {
  const { hasDefaultNotification, operationType, actionType, permissionForCreateLRON, permissionForViewLRON } = props;
  const permissionValue = getPermissionValue(permissionForViewLRON, permissionForCreateLRON, hasDefaultNotification);

  if (permissionValue === "110") {
    return null;
  }

  let title: React.ReactChild = <></>;
  switch (permissionValue) {
    case "000":
      title = (
        <>
          Your administrator may have set default notification settings for {OperationTypeMapTitle[operationType || actionType]} operations.
          Contact your administrator to request access.
        </>
      );
      break;
    case "010":
      title = (
        <>
          Your administrator may have set default notification settings for {OperationTypeMapTitle[operationType || actionType]} operations.
          You can send additional notifications for this operation.
        </>
      );
      break;
    case "100":
      title = (
        <>
          Your administrator has not set default notification settings for {OperationTypeMapTitle[operationType || actionType]} operations.
        </>
      );
      break;
    case "101":
      title = (
        <>
          Your administrator has set default notification settings for {OperationTypeMapTitle[operationType || actionType]} operations.
          Contact your administrator to request access.
        </>
      );
      break;
    case "111":
      title = (
        <>
          Default notification settings are set for {OperationTypeMapTitle[operationType || actionType]} operations. Configure default
          settings at{" "}
          <EuiLink style={{ textDecoration: "underline" }} href={`#${ROUTES.NOTIFICATIONS}`} target="_blank" external={false}>
            Notifications settings
          </EuiLink>
          .
        </>
      );
      break;
  }
  return (
    <>
      <EuiSpacer size="s" />
      <EuiCallOut color="primary" iconType="iInCircle" data-test-subj="defaultNotificationCallout" title={title} />
    </>
  );
}
