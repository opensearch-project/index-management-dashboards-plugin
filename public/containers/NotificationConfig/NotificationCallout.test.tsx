/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, waitFor } from "@testing-library/react";
import NotificationCallout, { INotificationCalloutProps } from "./NotificationCallout";
import { browserServicesMock } from "../../../test/mocks";
import { ActionType, OperationType } from "../../pages/Notifications/constant";
import { IAPICaller } from "../../../models/interfaces";

function renderNotification(props: Omit<INotificationCalloutProps, "actionType" | "operationType">) {
  return {
    ...render(<NotificationCallout actionType={ActionType.RESIZE} operationType={OperationType.SHRINK} {...props} />),
  };
}
/**
 * xyz
 * x has view permission
 * y has create permission
 * z has default notification
 * all the combos:
 * 000, 010, 100, 101, 110, 111
 */
describe("<ChannelNotification /> spec", () => {
  /**
   * 000
   */
  it("renders with no permission and no default notification", async () => {
    const { container, findByText } = renderNotification({
      permissionForViewLRON: false,
      permissionForCreateLRON: false,
      hasDefaultNotification: false,
    });
    await findByText(/Your administrator may have set default notification settings for/);
    expect(container).toMatchSnapshot();
  });

  /**
   * 010
   */
  it("renders with create permission and no default notification", async () => {
    const { container, findByText } = renderNotification({
      permissionForViewLRON: false,
      permissionForCreateLRON: true,
      hasDefaultNotification: false,
    });
    await findByText(/Your administrator may have set default notification settings for/);
    await findByText(/You can send additional notifications for this operation/);
    expect(container).toMatchSnapshot();
  });

  /**
   * 100
   */
  it("renders with view permission and no default notification", async () => {
    const { container } = renderNotification({
      permissionForViewLRON: true,
      permissionForCreateLRON: false,
      hasDefaultNotification: false,
    });
    expect(container).toMatchSnapshot();
  });

  /**
   * 101
   */
  it("renders with view permission and default notification", async () => {
    const { container, findByText } = renderNotification({
      permissionForViewLRON: true,
      permissionForCreateLRON: false,
      hasDefaultNotification: true,
    });
    await findByText(/Your administrator has set default notification settings for/);
    expect(container).toMatchSnapshot();
  });

  /**
   * 110
   */
  it("renders with full permission and no default notification", async () => {
    const { queryByTestId } = renderNotification({
      permissionForViewLRON: true,
      permissionForCreateLRON: true,
      hasDefaultNotification: false,
    });
    await waitFor(() => expect(queryByTestId("defaultNotificationCallout")).toBeNull());
  });

  /**
   * 111
   */
  it("renders with full permission and default notification", async () => {
    const { container, findByText } = renderNotification({
      permissionForViewLRON: true,
      permissionForCreateLRON: true,
      hasDefaultNotification: true,
    });
    await findByText(/Configure default settings at/);
    expect(container).toMatchSnapshot();
  });
});
