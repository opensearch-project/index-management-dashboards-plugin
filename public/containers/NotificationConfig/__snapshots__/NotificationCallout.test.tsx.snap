// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChannelNotification /> spec renders with create permission and no default notification 1`] = `
<div>
  <div
    class="euiSpacer euiSpacer--s"
  />
  <div
    class="euiCallOut euiCallOut--primary"
    data-test-subj="defaultNotificationCallout"
  >
    <div
      class="euiCallOutHeader"
    >
      EuiIconMock
      <span
        class="euiCallOutHeader__title"
      >
        Your administrator may have set default notification settings for 
        shrink
         operations. You can send additional notifications for this operation.
      </span>
    </div>
  </div>
</div>
`;

exports[`<ChannelNotification /> spec renders with full permission and default notification 1`] = `
<div>
  <div
    class="euiSpacer euiSpacer--s"
  />
  <div
    class="euiCallOut euiCallOut--primary"
    data-test-subj="defaultNotificationCallout"
  >
    <div
      class="euiCallOutHeader"
    >
      EuiIconMock
      <span
        class="euiCallOutHeader__title"
      >
        Default notification settings are set for 
        shrink
         operations. Configure default settings at
         
        <a
          class="euiLink euiLink--primary"
          href="#/notifications"
          rel="noopener noreferrer"
          style="text-decoration: underline;"
          target="_blank"
        >
          Notifications settings
          <span
            class="euiScreenReaderOnly"
          >
            (opens in a new tab or window)
          </span>
        </a>
        .
      </span>
    </div>
  </div>
</div>
`;

exports[`<ChannelNotification /> spec renders with no permission and no default notification 1`] = `
<div>
  <div
    class="euiSpacer euiSpacer--s"
  />
  <div
    class="euiCallOut euiCallOut--primary"
    data-test-subj="defaultNotificationCallout"
  >
    <div
      class="euiCallOutHeader"
    >
      EuiIconMock
      <span
        class="euiCallOutHeader__title"
      >
        Your administrator may have set default notification settings for 
        shrink
         operations. Contact your administrator to request access.
      </span>
    </div>
  </div>
</div>
`;

exports[`<ChannelNotification /> spec renders with view permission and default notification 1`] = `
<div>
  <div
    class="euiSpacer euiSpacer--s"
  />
  <div
    class="euiCallOut euiCallOut--primary"
    data-test-subj="defaultNotificationCallout"
  >
    <div
      class="euiCallOutHeader"
    >
      EuiIconMock
      <span
        class="euiCallOutHeader__title"
      >
        Your administrator has set default notification settings for 
        shrink
         operations. Contact your administrator to request access.
      </span>
    </div>
  </div>
</div>
`;

exports[`<ChannelNotification /> spec renders with view permission and no default notification 1`] = `
<div>
  <div
    class="euiSpacer euiSpacer--s"
  />
  <div
    class="euiCallOut euiCallOut--primary"
    data-test-subj="defaultNotificationCallout"
  >
    <div
      class="euiCallOutHeader"
    >
      EuiIconMock
      <span
        class="euiCallOutHeader__title"
      >
        Your administrator has not set default notification settings for 
        shrink
         operations.
      </span>
    </div>
  </div>
</div>
`;
